src/states/adult/mod.rs

```diff
@@ -321,118 +322,55 @@ impl Adult {
-            self.handle_signed_message(signed_msg)
+        if let Some(msg) = self.sig_accumulator.add_proof(msg) {
+            self.handle_accumulated_message(msg, outbox)
         } else {
             Ok(Transition::Stay)
         }
     }
 
-    // If the message is for us, verify it then, handle the enclosed routing message and swarm it
-    // to the rest of our section when destination is targeting multiple; if not, forward it.
-    fn handle_signed_message(
+    fn handle_accumulated_message(
         &mut self,
-        msg: HopMessageWithBytes,
-    ) -> Result<Transition, RoutingError> {
-        if !self.routing_msg_filter.filter_incoming(&msg).is_new() {
+        msg: Message,
+        outbox: &mut dyn EventBox,
+    ) -> Result<Transition> {
+        let mut msg = MessageWithBytes::new(msg, &self.log_ident())?;
+
+        if !self.filter_incoming_message(&msg) {
             trace!(
                 "{} Known message: {:?} - not handling further",
                 self,
-                msg.full_message_crypto_hash()
+                msg.full_crypto_hash()
             );
+
             return Ok(Transition::Stay);
         }
 
-        self.handle_filtered_signed_message(msg)
-    }
-
-    fn handle_filtered_signed_message(
-        &mut self,
-        mut msg: HopMessageWithBytes,
-    ) -> Result<Transition, RoutingError> {
-        trace!(
-            "{} - Handle signed message: {:?}",
-            self,
-            msg.full_message_crypto_hash()
-        );
-
-        if self.in_dst_location(msg.message_dst()) {
-            let signed_msg = msg.take_or_deserialize_signed_routing_message()?;
-            match &signed_msg.routing_message().content {
-                Variant::GenesisUpdate(info) => {
-                    self.verify_signed_message(&signed_msg)?;
-                    return self.handle_genesis_update((**info).clone());
-                }
-                Variant::Relocate(_) => {
-                    self.verify_signed_message(&signed_msg)?;
-                    let signed_relocate = SignedRelocateDetails::new(signed_msg.clone())?;
-                    return self.handle_relocate(signed_relocate);
-                }
-                _ => {
-                    self.msg_backlog.push(signed_msg.into_queued());
-                }
-            }
+        if self.should_relay_message(msg.message_dst()) {
+            self.relay_message(None, msg.clone())?;
         }
 
-        self.send_signed_message_to_elders(msg)?;
-        Ok(Transition::Stay)
-    }
-
-    fn verify_signed_message(&self, msg: &SignedRoutingMessage) -> Result<(), RoutingError> {
-        msg.verify(self.chain.get_their_key_infos())
-            .and_then(VerifyStatus::require_full)
-            .map_err(|error| {
-                self.log_verify_failure(msg, &error, self.chain.get_their_key_infos());
-                error
-            })
+        let msg = msg.take_or_deserialize_message()?;
+        if self.should_handle_message(&msg)? {
```

Comment:
That is not quite right, we should not `take_or_deserialize_message` if we do not have `should_handle_message` true, that defeat the purpose.

Reply:
Maybe an option would be to have a function like, that does the take_or_deserialize_message inside it (since you want to do the verify in it).

```
if let Some(msg) = self.should_handle_message(&msg)? {
     self.handle_message(None, msg, outbox)
}
```
Probably need a different name.